internal/scraper/job.go

```diff
@@ -0,0 +1,110 @@
+package scraper
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	discoveryv1 "github.com/syncromatics/kafmesh/internal/protos/kafmesh/discovery/v1"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//go:generate mockgen -source=./job.go -destination=./job_mock_test.go -package=scraper_test
+
+const (
+	scrapeAnnotation = "kafmesh/scrape"
+	portAnnotation   = "kafmesh/port"
+)
+
+// PodLister gets the pods running in the cluster
+type PodLister interface {
+	List(context.Context, metav1.ListOptions) (*v1.PodList, error)
+}
+
+// DiscoveryClient is the kafmesh discovery grpc client
+type DiscoveryClient interface {
+	GetServiceInfo(context.Context, *discoveryv1.GetServiceInfoRequest, ...grpc.CallOption) (*discoveryv1.GetServiceInfoResponse, error)
+}
+
+// DiscoveryFactory creates DiscoveryClients for pods
+type DiscoveryFactory interface {
+	Client(context.Context, string) (DiscoveryClient, func() error, error)
+}
+
+// Job runs scrape jobs against pods running in the cluster
+type Job struct {
+	podLister        PodLister
+	discoveryFactory DiscoveryFactory
+}
+
+// NewJob creates a new job
+func NewJob(podLister PodLister, discoveryFactory DiscoveryFactory) *Job {
+	return &Job{podLister, discoveryFactory}
+}
+
+// Scrape scrapes the kafmesh pods running in the kubernetes cluster
+func (j *Job) Scrape(ctx context.Context) (map[string]*discoveryv1.Service, error) {
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+
+	pods, err := j.getPods(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get pods")
+	}
+
+	results := map[string]*discoveryv1.Service{}
+
+	for _, pod := range pods {
+		r, err := j.scrapePod(ctx, pod)
+		if err != nil {
+			fmt.Printf(errors.Wrapf(err, "failed to scrape pod '%s'\n", pod.Name).Error())
```

Comment:
If scraping fails, this log message will be somewhat difficult to pick out. I think at a minimum, this should write to `os.Stderr` instead of the default `os.Stdout`.

Reply:
I agree. I think we should probably fail the whole job and log an error. A failing pod could lead to inconsistent results being saved to the datastore.

Reply:
oh i missed this one